### PR TITLE
DO NOT MERGE - [WIP,Experimental] - Add observable support.

### PR DESCRIPTION
**Pull request by jamestalmage**
Quick hack to make it work.

 Proof of concept for #20.

 I think I really like what this can do for testing

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -117,6 +131,38 @@ module.exports = function (cmd, args, opts) {
 		handleInput(spawned, parsed.opts);
 	});
 
+	if (opts && (opts.stdout === 'observable' || (opts.stdout && opts.stdout.observable))) {
```
Comment (jamestalmage) [inline index.js]: The options handling here is not well thought out. Just wanted to get something working.

File: test.js [hunk 2]
```diff
@@ -108,3 +109,26 @@ test('execa() returns a promise with kill() and pid', t => {
...
 });
+
+test('observable on stdin', t => {
+	t.plan(2);
+
+	return m('stdin', [], {stdout: 'observable', input: 'hello\ngoodbye\nhello'})
+		.stdout
+		.filter(line => /h/.test(line))
+		.map(line => line.toUpperCase())
+		.forEach(line => t.is(line, 'HELLO'));
```
Comment (jamestalmage) [inline test.js]: Overdone example here, just showing the power of `Observables`.

Understand that the `.filter`, `.map` and `.forEach` methods are not synchronous like Array methods. And they fire as lines are encountered. The second call to the `.filter` method could happen long after the first call to `.forEach`.

File: index.js [hunk 3]
```diff
@@ -5,7 +5,21 @@ var stripEof = require('strip-eof');
...
 var pathKey = require('path-key')();
+
+function tryRequireObservable() {
+	if (typeof global.Observable === 'function') {
+		return global.Observable;
+	}
+
+	try {
+		return require('zen-observable');
+	} catch (e) {
```
Comment (jamestalmage) [inline index.js]: - `global.Observable`
- `require('zen-observable')`
- `require('rxjs/Observable')`?

Which order?

**Comment by sindresorhus** [comment]
This will be awesome for testing! I do think we should finish the `execFile` => `spawn` stuff first though.

**Comment by jamestalmage** [comment]
Oh yes - this is throwaway now. I will redo using `stream-to-observable`.

I also want to land an in-progress option handling PR before this as well. (opts.stdin vs opts.stdio).

We need to rework option-handling to make this easy to use anyways.

**Comment by BarryThePenguin** [comment]
Curious how this will differ to [spawn-rx](/paulcbetts/spawn-rx)?

**Comment by sindresorhus** [comment]
Closing this as it hasn't seen any activity in a long time.

**Pull request closed without merging**